File: python/antenna_ml/materials.py
from typing import Dict, Optional
from .types import Material

DEFAULT_MATERIALS: Dict[str, Material] = {
    "PEC": Material(
        name="Perfect Electric Conductor",
        epsilon_r=1.0,
        mu_r=1.0,
        sigma=1e10,
        tan_delta=0.0
    ),
    "Copper": Material(
        name="Copper",
        epsilon_r=1.0,
        mu_r=1.0,
        sigma=5.96e7,
        tan_delta=0.0
    ),
    "FR4": Material(
        name="FR4 Substrate",
        epsilon_r=4.4,
        mu_r=1.0,
        sigma=0.0,
        tan_delta=0.02
    ),
    "Air": Material(
        name="Air",
        epsilon_r=1.0,
        mu_r=1.0,
        sigma=0.0,
        tan_delta=0.0
    ),
    "RT5880": Material(
        name="Rogers RT/duroid 5880",
        epsilon_r=2.2,
        mu_r=1.0,
        sigma=0.0,
        tan_delta=0.0009
    )
}

class MaterialDatabase:
    def __init__(self, materials: Optional[Dict[str, Material]] = None):
        self.materials = materials or DEFAULT_MATERIALS.copy()
    
    def get_material(self, name: str) -> Optional[Material]:
        return self.materials.get(name)
    
    def add_material(self, material: Material) -> None:
        self.materials[material.name] = material
    
    def list_materials(self) -> list[str]:
        return list(self.materials.keys())
    
    def get_all_materials(self) -> Dict[str, Material]:
        return self.materials.copy()

File: src/types/materials.ts
import { Material } from './antenna';

export interface MaterialDatabase {
  materials: Record<string, Material>;
}

export const DEFAULT_MATERIALS: Record<string, Material> = {
  PEC: {
    name: 'Perfect Electric Conductor',
    epsilonR: 1.0,
    muR: 1.0,
    sigma: 1e10,
    tanDelta: 0.0
  },
  Copper: {
    name: 'Copper',
    epsilonR: 1.0,
    muR: 1.0,
    sigma: 5.96e7,
    tanDelta: 0.0
  },
  FR4: {
    name: 'FR4 Substrate',
    epsilonR: 4.4,
    muR: 1.0,
    sigma: 0.0,
    tanDelta: 0.02
  },
  Air: {
    name: 'Air',
    epsilonR: 1.0,
    muR: 1.0,
    sigma: 0.0,
    tanDelta: 0.0
  },
  RT5880: {
    name: 'Rogers RT/duroid 5880',
    epsilonR: 2.2,
    muR: 1.0,
    sigma: 0.0,
    tanDelta: 0.0009
  }
};

export class MaterialManager {
  private materials: Record<string, Material>;
  
  constructor(initialMaterials: Record<string, Material> = DEFAULT_MATERIALS) {
    this.materials = { ...initialMaterials };
  }
  
  getMaterial(name: string): Material | undefined {
    return this.materials[name];
  }
  
  addMaterial(material: Material): void {
    this.materials[material.name] = material;
  }
  
  listMaterials(): string[] {
    return Object.keys(this.materials);
  }
  
  getAllMaterials(): Record<string, Material> {
    return { ...this.materials };
  }
}